src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -3780,6 +3782,12 @@ bool IdealLoopTree::iteration_split(PhaseIdealLoop* phase, Node_List &old_new) {
         phase->do_unswitching(this, old_new);
         return false; // need to recalculate idom data
       }
+      // if the loop body has a ScopedValueGetResult for a loop invariant ScopedValue object that dominates the backedge,
+      // then peeling one iteration of the loop body will allow the entire ScopedValue subgraph from the loop body
+      if (policy_peeling(phase, true)) {
+        phase->do_peeling(this, old_new);
+        return false;
+      }
```

Comment:
Just because I'm curious: why do the other places not already peel these loops? I.e. why do we need this here?

Reply:
Peeling looks for a loop invariant condition with one branch that exits the loop because then peeling makes the test in the loop body redundant with the one in the peeled iteration. Here, if there's a `ScopedValue.get()` on a loop invariant `ScopedValue` object, peeling one iteration will make `ScopedValue.get()` in the loop body redundant with the one in the peeled iteration. So it's not quite the same, at least, because for `ScopedValue.get()` the optimization applies whether `ScopedValue.get()` causes an exit of the loop or not.

Reply:
Ah, great, thanks for the explanation!